stdlib/src/collections/list.mojo

```diff
@@ -467,38 +468,90 @@ struct List[T: CollectionElement, hint_trivial_type: Bool = False](
-            The bytecount of the List.
+            The byte length of the List.
         """
         return len(self) * sizeof[T]()
 
     fn _realloc(mut self, new_capacity: Int):
         var new_data = UnsafePointer[T].alloc(new_capacity)
 
-        _move_pointee_into_many_elements[hint_trivial_type](
-            dest=new_data,
-            src=self.data,
-            size=self.size,
-        )
+        @parameter
+        if hint_trivial_type:
+            memcpy(new_data, self.data, len(self))
+        else:
+            for i in range(len(self)):
+                (self.data + i).move_pointee_into(new_data + i)
 
         if self.data:
             self.data.free()
         self.data = new_data
         self.capacity = new_capacity
 
     fn append(mut self, owned value: T):
-        """Appends a value to this list.
+        """Appends a value to this list. If there is no capacity left, resizes
+        to twice the current capacity. Except for 0 capacity where it sets 1.
 
         Args:
             value: The value to append.
         """
-        if self.size >= self.capacity:
-            self._realloc(max(1, self.capacity * 2))
-        (self.data + self.size).init_pointee_move(value^)
-        self.size += 1
+        if len(self) >= self.capacity:
+            self._realloc(self.capacity * 2 + int(self.capacity == 0))
+        (self.data + len(self)).init_pointee_move(value^)
+        self._len += 1
+
+    fn append[
+        D: DType, //
+    ](mut self: List[Scalar[D], *_, **_], value: SIMD[D, _]):
+        """Appends a vector to this list. If there is no capacity left, resizes
+        to `len(self) + value.size`.
+
+        Parameters:
+            D: The DType.
+
+        Args:
+            value: The value to append.
+        """
+        self.reserve(len(self) + value.size)
+        (self.data + len(self)).store(value)
+        self._len += value.size
+
+    fn append[
+        D: DType, //
+    ](mut self: List[Scalar[D], *_, **_], value: SIMD[D, _], count: Int):
+        """Appends a vector to this list. If there is no capacity left, resizes
+        to `len(self) + count`.
+
+        Parameters:
+            D: The DType.
+
+        Args:
+            value: The value to append.
+            count: The ammount of items to append.
+        """
+        self.reserve(len(self) + count)
+        var v_ptr = UnsafePointer.address_of(value).bitcast[Scalar[D]]()
+        memcpy(self.data + len(self), v_ptr, count)
+        self._len += count
+
+    fn append[
+        D: DType, //
+    ](mut self: List[Scalar[D], *_, **_], value: Span[Scalar[D]]):
+        """Appends a Span to this list. If there is no capacity left, resizes
+        to `len(self) + len(value)`.
+
+        Parameters:
+            D: The DType.
+
+        Args:
+            value: The value to append.
+        """
+        self.reserve(len(self) + len(value))
+        memcpy(self.data + len(self), value.unsafe_ptr(), len(value))
+        self._len += len(value)
```

Comment:
**Suggestion** I think it would be better if we leave these new `append()` overloads to a separate PR. I know it can feel like a small thing to add QoI "goodies" to a PR like this, but it can make the PR significantly harder to review when its changing multiple things at once 😌 

Layering multiple changes in one PR can feel more efficient since it means more changes can merge "all at once", but I think it might lower overall throughput since each PR takes more brain 🧠 to review.

For some additional context, when we're reviewing community PRs, it's often in between other unrelated internal work. A small PR can often be reviewed in a few small gaps throughout the day, since small means minimum context needs to get loaded into "working memory" 🧠, means more potential time slots are big enough to make progress on the review.

On the other hand, a larger PR might not fit as easily, which unfortunately means longer reviews and longer delays. It can be especially inefficient if that delay holds up reviews of subsequent smaller PRs that might have been able to be reviewed earlier if they weren't "stuck" behind a larger PR.

We love your work Martin ♥️ (and the work of all of our community members), and we'll always do our best to review your PRs as efficiently and quickly as we can while maintaining our review quality standards. I just wanted to give some context and suggestions on how we might all benefit from small, targeted PRs 🙂

Reply:
ok, will do. I added these here because splitting them off would mean making this PR a draft until the other one is merged, since the functionality needs to edit `List.size`.

PS: just tell me when you'd like something to be split off, I also want to help you guys makes the whole endeavor faster. I'm still learning to contribute to OSS, I'm used to having merge access and making monster changes :rofl:.

Reply:
Thanks Martin 🙂

> I added these here because splitting them off would mean making this PR a draft until the other one is merged, since the functionality needs to edit List.size.

Am I understanding correctly that's because of visibility? If so, I think that other PR could be decoupled from this one by mutating `list.size` / `list._len` directly. `_len` makes a field "notionally" / "conventionally" private, but its still okay in special circumstances to mutate it directly 🙂  It just should be done with care.

> PS: just tell me when you'd like something to be split off, I also want to help you guys makes the whole endeavor faster. I'm still learning to contribute to OSS, I'm used to having merge access and making monster changes 🤣.

Thanks Martin, we're all learning together here 🙂  It's been an adjustment for me as well; in the past I've mostly worked on codebases where I was the sole contributor and could make sweeping changes quickly. Here's to us learning to make smaller changes and learning how to use the process most effectively 🍻 🙂 

I think a good rule of thumb is PRs should generally try to do only one thing, e.g. fix a bug, perform a refactoring, or add a new feature, but maybe not multiple in the same PR. That helps with reading the PR. E.g. if I know a PR is only supposed to be doing a refactoring, I'll read it with a different mindset than if I know its adding new (potentially unproved) code.